envision/web/src/components/simulation.js

```diff
@@ -149,19 +149,25 @@ export default function Simulation({
     setScene(scene_);
   };
 
+  const sleep = (milliseconds) => {
+    return new Promise((resolve) => setTimeout(resolve, milliseconds));
+  };
+
   // State subscription
   useEffect(() => {
     let stopPolling = false;
     (async () => {
       const it = client.worldstate(simulationId);
-      let wstate_and_time = await it.next();
-      while (!wstate_and_time.done && playing) {
+      let wstate_and_time;
+      if (playing) wstate_and_time = await it.next();
+      while (!stopPolling && playing && !wstate_and_time.done) {
         let wstate, elapsed_times;
         [wstate, elapsed_times] = wstate_and_time.value;
-        if (!stopPolling) {
-          setWorldState(wstate);
-          onElapsedTimesChanged(...elapsed_times);
-        }
+        setWorldState(wstate);
+        onElapsedTimesChanged(...elapsed_times);
+
+        // play the frames every 50ms
+        await sleep(50);
```

Comment:
Will try to use the difference between the elapsed time of previous frame and current frame to decide how long to sleep. Also needs to take into account the time spent on waiting to get the next frame